### ISSUE TITLE
Excessive DeepCopy usage in case of using remote state backend

### ISSUE THREAD

**Issue by alexott** [opentofu/opentofu#1580]
### OpenTofu Version

```shell
main branch
```

### OpenTofu Configuration Files

...

### Debug Output

...

### Expected Behavior

We should avoid doing unnecessary `DeepCopy` on the state

### Actual Behavior

When persisting the state, we're calling `DeepCopy` on state in the [internal/tofu/update_state_hook.go](https://github.com/opentofu/opentofu/blob/main/internal/tofu/update_state_hook.go#L16) and then in the [remote state backend](https://github.com/opentofu/opentofu/blob/main/internal/states/remote/state.go#L107), and with big states this causes significant slowness.

### Steps to Reproduce

Have a terraform code with a lot of resources (10ks), run apply with remote state backend configured.If you have Terraform code with a lot of resources (10ks), run apply with the

**Comment by alexott** [comment]
I can share the Terraform code (it's synthetic anyway), but it uses Databricks resources (directories, notebooks, etc.). But it requires a Databricks workspace to test against (it shouldn't incur costs as it's charged for compute and data, and notebooks are "free").

Let me talk internally about testbed

**Comment by alexott** [comment]
Potentially, we can use the free tier of Cloudflare R2 (via AWS TF provider) by creating 100k-300k tiny blobs in it... The main performance problems come from the number of resources, not necessary from dependencies.

**Comment by cam72cam** [comment]
I think if the example uses the tfcoremock provider, we wouldn't have to worry about credentials / other overhead?

Most of the performance testing you have been doing is focused around overhead in tofu, not the individual providers.

**Comment by alexott** [comment]
Yes, I think so. We are fixing performance issues in our provider as well, but we hit issues in the tofu as well.

**Comment by cam72cam** [comment]
Linking to my comment on the above PR: https://github.com/opentofu/opentofu/pull/1598#issuecomment-2120780580

I think that we can look at hot-spots in the code that call deep copy multiple times and instead require an immutable state passed in instead.  It's then up to the caller to provider an immutable state and will allow us to optimize the paths that currently require multiple DeepCopy calls.  We should be able to slowly replace most usages of DeepCopy with an immutable state and reduce DeepCopy calls as we move our way through calling code paths.

An approach in this vein adds a lot of compile time guarantees that are currently assumptions / overzealous uses of DeepCopy.

**Comment by cam72cam** [comment]
I've started hacking and slashing my way through the code adding ImmutableState.  It's a complete tangled mess of DeepCopy sprinkled anywhere someone wasn't sure if it was needed or not.  I'll push a WIP branch later today so we have a concrete idea of what this would look like.  Honestly, it'll probably fix a few bugs deep in the codebase.

**Comment by alexott** [comment]
@cam72cam wow! thank you for digging into it!

**Comment by cam72cam** [comment]
You can take a look at the branch here:
https://github.com/opentofu/opentofu/tree/1580_immutable_state_prototype
Note: tests broken, this is just a hack to see the impact.

I think long term this is the right move.  We should work toward making state immutable wherever possible through the entire application.  There's a few cases that use states.SyncWrapper that truly are dynamic.  The vast majority of state writes happen through that struct.

**Comment by cam72cam** [comment]
I've added myself as an additional assignee so I can write up my findings about immutable state and the risks currently present in the codebase.

**Comment by cam72cam** [comment]
I'd like to submit that we accept this issue, with #3011 as the solution, barring further refactoring.

It reduces the execution of [our standard test](https://github.com/opentofu/opentofu-performance-testing/tree/main/wide-graph-with-submodules) from 6m23.078s to 2m56.235s, which is also shown in the profile

<img width="2557" height="710" alt="Image" src="https://github.com/user-attachments/assets/442b99c2-a8c6-4882-99b3-36916f142312" />

<img width="2557" height="710" alt="Image" src="https://github.com/user-attachments/assets/6028cf95-c986-4c0c-958c-3fd612ee3f58" />

Not shown is the GC time down from 540s to 249s due to object churn)


This is a large improvement, and I think the furthest we can go barring more fundamental changes to our state storage interface.  There's a remaining large DeepCopy in there which is not ideal.

**Comment by cam72cam** [comment]
I've also pushed some changes to https://github.com/opentofu/opentofu/tree/1580_aggressive_optimization which bring the total run time for this test on my system to about 42s

<img width="2557" height="710" alt="Image" src="https://github.com/user-attachments/assets/34726824-7b01-43ed-bb17-cf3320022c67" />

I'll be PRing that separately as it's a whole extra level of invasive change.

**Comment by cam72cam** [comment]
If this is accepted, the results will be documented in opentofu/opentofu-performance-testing.

**Comment by cam72cam** [comment]
Had some conversations with the other Maintainers.  Marking this as accepted and am working on getting both PRs ready to review.

**Comment by alexott** [comment]
Almost 13x speedup! This is huge
